### PR TITLE
HHH-15848 Session.isDirty() and more (rebase)

### PR DESCRIPTION
**Pull request by gavinking**
Supersedes #9379

[Please describe here what your change is about]

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-15848

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java [hunk 1]
```diff
@@ -357,49 +357,54 @@ public boolean requiresDirtyCheck(Object entity) {
...
+	}
+
+	private boolean isNonDirtyViaCustomStrategy(Object entity) {
+		if ( isPersistentAttributeInterceptable( entity ) ) {
+			final PersistentAttributeInterceptor interceptor =
+					asPersistentAttributeInterceptable( entity ).$$_hibernate_getInterceptor();
+			if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
+				// we never have to check an uninitialized proxy
+				// TODO: why do we not check !lazinessInterceptor.hasWrittenFieldNames()
+				//       as we do below in isNonDirtyViaTracker() ?
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java:372]: @dreab8 This is a question for you.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java:372]: @gavinking  with dirty checking  setting a property will not trigger the initialization, while without dirty checking setting a property triggers the enhamced proxy initilaization and then the interceptor would not be of type `EnhancementAsProxyLazinessInterceptor`

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java:372]: basically `writtenFieldNames` is populated only when dirty checking is enabled

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java:372]: @dreab8 I guess I don't understand what is the difference between the two branches. They both do:

```java
if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
```

and then one of them does the additional check, and one doesn't.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java:372]: @gavinking one branch deals wit the case dirty checking is not enabled, in this case it's enough `interceptor instanceof EnhancementAsProxyLazinessInterceptor` to be sure it's not dirty, the other branch deals with the case dirty checking is enabled and if `interceptor instanceof EnhancementAsProxyLazinessInterceptor` we need also to be sure it has not written field names to say that it's not dirty

**Pull request merged**